src/tbb/arena.h

```diff
@@ -179,6 +179,82 @@ class atomic_flag {
     }
 };
 
+#if __TBB_PREVIEW_PARALLEL_PHASE
+class thread_leave_manager {
+    static const std::uintptr_t DELAYED_LEAVE       = 0;
+    static const std::uintptr_t FAST_LEAVE          = 1;
+    static const std::uintptr_t ONE_TIME_FAST_LEAVE = 1 << 1;
+    static const std::uintptr_t PARALLEL_PHASE      = 1 << 2;
+
+    std::atomic<std::uintptr_t> my_state{static_cast<std::uintptr_t>(-1)};
+public:
+    // This method is not thread-safe!
+    // Required to be called after construction to set initial state of the state machine.
+    void set_initial_state(tbb::task_arena::leave_policy lp) {
+        if (lp == tbb::task_arena::leave_policy::automatic) {
+            std::uintptr_t platform_policy = governor::hybrid_cpu() ? FAST_LEAVE : DELAYED_LEAVE;
+            my_state.store(platform_policy, std::memory_order_relaxed);
+        } else {
+            __TBB_ASSERT(lp == tbb::task_arena::leave_policy::fast,
+                         "Was the new value introduced for leave policy?");
+            my_state.store(FAST_LEAVE, std::memory_order_relaxed);
+        }
+    }
+
+    void reset_if_needed() {
+        std::uintptr_t curr = ONE_TIME_FAST_LEAVE;
+        if (my_state.load(std::memory_order_relaxed) == curr) {
+            // Potentially can override decision of the parallel block from future epoch
+            // but it is not a problem because it does not violate the correctness
+            my_state.compare_exchange_strong(curr, DELAYED_LEAVE);
+        }
+    }
+
+    // Indicate start of parallel phase in the state machine
+    void register_parallel_phase() {
+        std::uintptr_t prev = my_state.load(std::memory_order_relaxed);
+        __TBB_ASSERT(prev != std::uintptr_t(-1), "The initial state was not set");
+
+        std::uintptr_t desired{};
+        do {
+            // Need to add a reference for this start of a parallel phase, preserving the leave
+            // policy. Except for the case when one time fast leave was requested at the end of a
+            // previous phase.
+            desired = prev;
+            if (prev == ONE_TIME_FAST_LEAVE) {
+                // State was previously transitioned to "One-time Fast leave", thus with the start
+                // of new parallel phase, it should be transitioned to "Delayed leave"
+                desired = DELAYED_LEAVE;
+            }
+            __TBB_ASSERT(desired + PARALLEL_PHASE > desired, "Overflow detected");
+            desired += PARALLEL_PHASE; // Take into account this start of a parallel phase
+        } while (!my_state.compare_exchange_strong(prev, desired));
```

Comment:
I think it can be simplified as `fetch_add(PARALLEL_PHASE)` followed by `fetch_and(~ONE_TIME_FAST_LEAVE)` in case the bit is set in the returned value of `fetch_add`.

Reply:
Done.